File: go/222.go
package main

import "math"

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func hr(t *TreeNode) int {
	var ret int = 0
	for ; t != nil; t = t.Right {
		ret++
	}
	return ret
}

func hl(t *TreeNode) int {
	var ret int = 0
	for ; t != nil; t = t.Left {
		ret++
	}
	return ret
}

func sol(t *TreeNode) int {
	var l, r int = hl(t), hr(t)
	if l == r { // full BT
		return int(math.Pow(2, float64(l))) - 1
	}
	return 1 + sol(t.Left) + sol(t.Right) // not full, so keep doing
}

func countNodes(root *TreeNode) int {
	return sol(root)
}

File: cpp/222.cpp
#include <iostream>
#include <queue>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
public:
  int countNodes(TreeNode *root) {
    if (!root)
      return 0;
    int ret = 0;
    queue<TreeNode *> q;
    q.push(root);
    while (!q.empty()) {
      int size = q.size();
      ret += size;
      for (; size > 0; size--) {
        if (q.front()->left)
          q.push(q.front()->left);
        if (q.front()->right)
          q.push(q.front()->right);
        q.pop();
      }
    }
    return ret;
  }
};
